viamonvif/device/device_test.go

```diff
@@ -0,0 +1,79 @@
+package device
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"go.viam.com/rdk/logging"
+	"go.viam.com/test"
+)
+
+func TestSendSoapNoHang(t *testing.T) {
+	logger := logging.NewTestLogger(t)
+
+	t.Run("context cancellation works", func(t *testing.T) {
+		// Channel to coordinate server shutdown
+		done := make(chan struct{})
+		// A server that will hang forever
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// For GetCapabilities request during initialization, we must return a valid response
+			if strings.Contains(r.Header.Get("Content-Type"), "soap") &&
+				strings.Contains(readBody(r), "GetCapabilities") {
+				w.Write([]byte(`<?xml version="1.0" encoding="UTF-8"?>
+					<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope">
+						<SOAP-ENV:Body>
+							<GetCapabilitiesResponse>
+								<Capabilities>
+									<Media>
+										<XAddr>http://example.com/onvif/media</XAddr>
+									</Media>
+									<PTZ>
+										<XAddr>http://example.com/onvif/ptz</XAddr>
+									</PTZ>
+								</Capabilities>
+							</GetCapabilitiesResponse>
+						</SOAP-ENV:Body>
+					</SOAP-ENV:Envelope>`))
+				return
+			}
+			<-done
+		}))
+		defer func() {
+			close(done)
+			server.Close()
+		}()
+
+		serverURL, err := url.Parse(server.URL)
+		test.That(t, err, test.ShouldBeNil)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+
+		// Create device with the context
+		dev, err := NewDevice(Params{
+			Xaddr:      serverURL,
+			HTTPClient: &http.Client{},
+			Context:    ctx,
+		}, logger)
+		test.That(t, err, test.ShouldBeNil)
+
+		_, err = dev.sendSoap(server.URL, "test message")
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "context deadline exceeded")
```

Comment:
Test using test.ShouldBeError not substring matching

Reply:
Done. Had to cast the outer error to a url error though